overlord/configstate/configcore/prompting.go

```diff
@@ -0,0 +1,66 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+//go:build !nomanagers
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package configcore
+
+import (
+	"github.com/snapcore/snapd/features"
+	"github.com/snapcore/snapd/overlord/configstate/config"
+	"github.com/snapcore/snapd/overlord/restart"
+)
+
+// Trigger a security profile regeneration by restarting snapd if a change in
+// the experimental apparmor-prompting flag causes a need to do so.
+func doExperimentalApparmorPromptingDaemonRestart(c RunTransaction, opts *fsOnlyContext) error {
+	st := c.State()
+
+	snap, confName := features.AppArmorPrompting.ConfigOption()
+
+	var prompting bool
+	err := c.Get(snap, confName, &prompting)
+	if err != nil && !config.IsNoOption(err) {
+		return err
+	}
+	var prevPrompting bool
+	err = c.GetPristine(snap, confName, &prevPrompting)
+	if err != nil && !config.IsNoOption(err) {
+		return err
+	}
+	if prompting == prevPrompting {
+		return nil
+	}
+
+	// XXX: what if apparmor-prompting flag value unchanged but support changed?
+	// Prompting support (currently) depends only on AppArmor kernel and parser
+	// features, which are checked only once at startup, so this cannot occur.
+
+	if !features.AppArmorPrompting.IsSupported() {
```

Comment:
AFAIK, `IsSupported()` calls `apparmor.KernelFeatures()` and `apparmor.ParserFeatures()`, which in turn call `apparmor.probeKernelFeatures()` and `apparmor.probeParserFeatures()` using `sync.Once` to only probe once. These are evaluated as part of snapd initial startup, and any subsequent calls use the cached features initially evaluated. The kernel and parser features are both part of the system key, as is the prompting setting. If apparmor is updated while snapd is running, a change in prompting support will be handled the same as any other change in apparmor support.

That is, if apparmor feature support changes, I believe snapd handles this correctly and reloads profiles, at the very least, and updates the list of features -- it might restart snapd as a whole. If this isn't the case, then prompting wouldn't be any more broken than other features of snapd which depend on apparmor features. I'd bet @zyga knows more about this than I do.

Reply:
Snapd does not reload anything unless interfaces are connected/disconnect, and even then the parser features/kernel features set remains unchanged, as you correctly observed they are probed during startup and cached. This will be correct as long as we're using vendored apparmor since it's already coupled with a particular snapd revision, but the distro apparmor is outside of snapd's control. TBH for consistency I would simply request a restart and not second guess anything.